sdk/typescript/runtime/main.go

```diff
@@ -131,79 +137,32 @@ func (t *TypescriptSdk) CodegenBase(ctx context.Context, modSource *ModuleSource
-		WithExec([]string{
-			codegenBinPath,
-			"--lang", "typescript",
-			"--output", ModSourceDirPath,
-			"--module-name", name,
-			"--module-context-path", filepath.Join(ModSourceDirPath, subPath),
-			"--introspection-json-path", schemaPath,
-		}, ContainerWithExecOpts{
-			ExperimentalPrivilegedNesting: true,
-		}).
-		Directory(filepath.Join(ModSourceDirPath, subPath, GenDir))
+	// Get a directory with the SDK sources installed and the generated client.
+	sdk := t.
+		installedSDK(base, detectedRuntime).
+		WithDirectory(".", t.generateClient(base, name, introspectionJSON, subPath))
 
 	base = base.
-		// Add codegen binary into a special path
-		WithMountedFile(codegenBinPath, t.SDKSourceDir.File("/codegen")).
 		// Add template directory
 		WithMountedDirectory("/opt/module", dag.CurrentModule().Source().Directory(".")).
-		// Mount users' module
-		WithMountedDirectory(ModSourceDirPath, modSource.ContextDirectory()).
-		WithWorkdir(filepath.Join(ModSourceDirPath, subPath)).
-		// Add sdk source code
-		WithDirectory(GenDir, t.SDKSourceDir, ContainerWithDirectoryOpts{
-			Exclude: []string{"codegen", "runtime"},
-		}).
-		// Add generated code
-		WithDirectory(GenDir, gen)
-
-	switch detectedRuntime {
-	case Bun:
-		base = base.
-			// Check if the project has existing source:
-			// if it does: add sdk as dev dependency
-			// if not: copy the template and replace QuickStart with the module name
-			WithExec([]string{"sh", "-c",
-				"if [ -f package.json ]; then  bun install ./sdk  --dev  && bun /opt/module/bin/__tsconfig.updator.ts; else cp -r /opt/module/template/*.json .; fi",
-			})
-	case Node:
-		base = base.
-			// Check if the project has existing source:
-			// if it does: add sdk as dev dependency
-			// if not: copy the template and replace QuickStart with the module name
-			WithExec([]string{"sh", "-c",
-				"if [ -f package.json ]; then  npm install --package-lock-only ./sdk  --dev  && tsx /opt/module/bin/__tsconfig.updator.ts; else cp -r /opt/module/template/*.json .; fi",
-			})
-	default:
-		return nil, fmt.Errorf("unknown runtime: %s", detectedRuntime)
-	}
-
-	return base.
-		// Check if there's a src directory with .ts files in it.
-		// If not, add the template file and replace QuickStart with the module name
-		// This cover the case where there's a package.json but no src directory.
-		WithExec([]string{"sh", "-c",
-			fmt.Sprintf("mkdir -p src && if ls src/*.ts >/dev/null 2>&1; then true; else cp /opt/module/template/src/index.ts src/index.ts && sed -i -e 's/QuickStart/%s/g' ./src/index.ts; fi", strcase.ToCamel(name))},
-		), nil
+		// Mount users' module with SDK sources and generated client in it.
+		WithMountedDirectory(ModSourceDirPath, modSource.
+			ContextDirectory().
+			WithDirectory(filepath.Join(subPath, GenDir), sdk),
+		).
+		WithWorkdir(filepath.Join(ModSourceDirPath, subPath))
+
+	return t.setupModule(ctx, base, detectedRuntime, name)
```

Comment:
Good refactor here :tada:

This was *very* hard to understand before, super glad to see this cleaned up! :heart: